niceman/distributions/debian.py

```diff
@@ -584,28 +564,17 @@ def _pkg_name_to_dpkg_list_file(name):
         return query
 
     def _get_pkgs_versions_and_sources(self, pkg_dicts):
-        # Define a function to process batches of packages and return
-        # a dictionary of parsed apt-cache policy output
-        def proc_apt_cache_policy_cmd(subqueries, cmd_results):
-            try:
-                out, _ = self._session.execute_command(
-                    ['apt-cache', 'policy'] + subqueries
-                )
-                out = utils.to_unicode(out, "utf-8")
-                cmd_results.update(parse_apt_cache_policy_pkgs_output(out))
-            except CommandError as _:
-                raise  # Currently raise any exception we encounter
-            return cmd_results
-
         # Convert package names to name:arch format
-        queries = []
-        for p in pkg_dicts:
-            queries.append("%(name)s:%(architecture)s" % p)
-        # Determine how big our batches can be
-        num_files = get_cmd_batch_len(queries, 16)
-        # Now process our file list and return the parsed data dict
-        results = batch_process_list(proc_apt_cache_policy_cmd, queries,
-                                     num_files, {})
+        queries = ["%(name)s:%(architecture)s" % p for p in pkg_dicts]
+        # Call apt-cache policy in batches
+        exec_gen = execute_command_batch(self._session,
+                                         ['apt-cache', 'policy'],
+                                         queries)
+        # Parse results into a single generator
+        results = (parse_apt_cache_policy_pkgs_output(out)
+                   for (out, _, _) in exec_gen)
+        # Combine sequence of dicts
+        results = {k: v for d in results for k, v in d.items()}
```

Comment:
IMHO it is worth a separate helper function with a description/test of the expected operation when (for some reason) the same `k` happens in multiple entries.  ATM it just replaces with a "new" value.
In our case here it would be unexpected and should throw an Exception (e.g., a `RuntimeError("Have obtained a new value %r for %r where previous known was %r" % (v, k, results[k]))`).  I know that it shouldn't happen, but again -- those damn computers some times raise a fight against us and we better do not miss their sneaky spies (probably I am reading too much of sci-fi recently, currently Avogadro Corp trilogy)

Reply:
Done.  (And trust no one. :)